src/protections/BanInviteRetraction.ts

```diff
@@ -0,0 +1,160 @@
+// SPDX-FileCopyrightText: 2025 Emma [it/its] @ Rory& <matrix:u/emma:rory.gay>
+//
+// SPDX-License-Identifier: AFL-3.0
+
+import {
+  AbstractProtection,
+  ActionResult,
+  CapabilitySet,
+  MembershipChange,
+  MembershipChangeType,
+  Ok,
+  PolicyListRevision,
+  PolicyRule,
+  PolicyRuleChange,
+  PolicyRuleType,
+  PowerLevelPermission,
+  ProtectedRoomsSet,
+  Protection,
+  ProtectionDescription,
+  Recommendation,
+  RoomMembershipRevision,
+  SimpleChangeType,
+  Task,
+  describeProtection,
+} from "matrix-protection-suite";
+import { Draupnir } from "../Draupnir";
+import { redactUserMessagesIn } from "../utils";
+import {
+  StringRoomID,
+  StringUserID,
+} from "@the-draupnir-project/matrix-basic-types";
+
+type BanInviteRetractionProtectionDescription = ProtectionDescription<Draupnir>;
+
+export class BanInviteRetractionProtection
+  extends AbstractProtection<BanInviteRetractionProtectionDescription>
+  implements Protection<BanInviteRetractionProtectionDescription> {
+  public constructor(
+    description: BanInviteRetractionProtectionDescription,
+    capabilities: CapabilitySet,
+    protectedRoomsSet: ProtectedRoomsSet,
+    private readonly draupnir: Draupnir,
+  ) {
+    super(description, capabilities, protectedRoomsSet, {
+      requiredPermissions: [PowerLevelPermission.Kick],
+    });
+  }
+
+  public retractInvitesForNewUserPolicy(policy: PolicyRule): void {
+    const rooms: StringRoomID[] = [];
+    if (policy.isGlob()) {
+      this.protectedRoomsSet.allProtectedRooms.forEach((room) =>
+        rooms.push(room.toRoomIDOrAlias()),
+      );
+    } else {
+      for (const roomMembership of this.protectedRoomsSet.setRoomMembership
+        .allRooms) {
+        const membership = roomMembership.membershipForUser(
+          policy.entity as StringUserID,
+        );
+        if (membership !== undefined) {
+          rooms.push(roomMembership.room.toRoomIDOrAlias());
+        }
+      }
+    }
+    void Task(
+      redactUserMessagesIn(
+        this.draupnir.client,
+        this.draupnir.managementRoomOutput,
+        policy.entity,
+        rooms,
+      ),
+    );
```

Comment:
Why do you want to do this in this protection and not allow redaction synchronisation protection handle it?

Reply:
Redacting the messages won't remove the invites btw? i don't think? if that's the intent

Reply:
Uh, did i miss this? it's supposed to be calling retractPendingInvites or whaterver